### ISSUE TITLE
Propose to combine SolutionUserObject into SolutionFunction

### ISSUE THREAD

**Issue by YaqiWang** [idaholab/moose#10126]
## Rationale
Currently, user have to create a `SolutionUserObject` separately for a `SolutionFunction`. If we can create the user object inside the solution function, user interface can be simplified.

## Description
 It is more intuitive for the user to see all the parameters in one place. `SolutionUserObject` does not have to be removed.

## Impact
Better user experience.

**Comment by dschwen** [comment]
But a single solution UO can serve multiple functions!

**Comment by YaqiWang** [comment]
That can be still supported. At least in our use cases, it is always one to one and splitting them into two objects does not make much sense.

**Comment by dschwen** [comment]
I don't understand that comment. The use case for us is reading in a big exodus file out of which we need multiple variables. With one single UO the file is read just once and then we add multiple functions to pipe the data through to an IC.

**Comment by andrsd** [comment]
I agree with @dschwen, single user object can provide data for multiple functions, so we want the exodus file load just once.

@YaqiWang Did you consider making an action that would do this for you?

**Comment by YaqiWang** [comment]
I did consider an action, but that will require a particular input syntax, which will be different from `Functions/*`. I was in mind to make `params.addRequiredParam<UserObjectName>("solution"` optional, when it is invalid, we will construct a `SolutionUserObject` inside `SolutionFunction` constructor and immediately associate it with the function.

**Comment by YaqiWang** [comment]
BTW, I do not quite understand why if there is any `SolutionUserObject`, we cannot use distributed mesh. We are not asking the mesh on solution user object to be distributed.

**Comment by permcody** [comment]
It probably just wasn't designed with that in mind. I'm sure it can be modernized. We still have several objects in the repo that are restricted until they can be looked at. Often it's just an assumption that developers make when building these for some quick simulation.

**Comment by andrsd** [comment]
> I did consider an action, but that will require a particular input syntax

True, but if you were to build `SolutionUserObject` inside `SolutionFunction` you would have to somehow supply parameters for that `SolutionUserObject`. So it is not just a simple matter of relaxing the requirements on the `"solution"` parameter in `SolutionFunction`. The moment you start injecting `SolutionUserObject` parameters into `SolutionFunction`, it is a bad design and it will be rejected.

Another approach might be to build a new `SolutionFunction`-derived class that would combine the user object and the function. There is a problem though that `solution` parameter would still be required. I do not think we can change required parameters into non-required ones in child classes. And replicating `SolutionFunction` is a no-go either...

**Comment by YaqiWang** [comment]
That is actually what I want, putting all parameters into one place. The derived class can suppress the parameter. Do you think a derived class is a good way?

**Comment by andrsd** [comment]
It is a better solution than changing `SolutionFunction`.  The code duplication problem is not there and the syntax will look like a Function. It would be fine with me.

**Comment by dschwen** [comment]
So we'd build a userobject on the add_function task?!

**Comment by YaqiWang** [comment]
Yeah. userobject does not have to be added by add_user_object task. We violate this kind of rules places I believe. I guess it is fine that we know nobody will be using the object and the object depends on nothing and the object name is carefully designed so that chance of naming conflict is small.

**Comment by permcody** [comment]
Ugh... I really don't like that people take advantage of these loopholes. It makes it difficult to refactor things later. I'm going to vote no against anything that relies on a hack. If we want to do something like this, we need to create the proper task for it and support it.

**Comment by YaqiWang** [comment]
Is it possible to create a utility class, which can be used by both SolutionUserObject and the combined solution function?

**Comment by permcody** [comment]
Yes, that's a better solution. I'm pretty lazy, but I thought that's what we were already doing (e.g. Isn't SolutionUserObject used by SolutionUserFunction?)

**Comment by YaqiWang** [comment]
Yes, but SolutionUserObject is a user object. I thought you do not want a function to construct a user object in its constructor.
